p2p/p2p.go

```diff
@@ -219,25 +219,28 @@ func (p *P2P) Dial(address *lib.PeerAddress, disconnect, strictPublicKey bool) l
 		return ErrFailedDial(er)
 	}
 	// try to use the basic tcp connection to establish a peer
-	return p.AddPeer(conn, &lib.PeerInfo{Address: address, IsOutbound: true}, disconnect, strictPublicKey)
+	return p.AddPeer(conn, &lib.PeerInfo{Address: address, IsOutbound: true}, disconnect, strictPublicKey, closeOnFail)
 }
 
 // AddPeer() takes an ephemeral tcp connection and an incomplete peerInfo and attempts to
 // create a E2E encrypted channel with a fully authenticated peer and save it to
 // the peer set and the peer book
-func (p *P2P) AddPeer(conn net.Conn, info *lib.PeerInfo, disconnect, strictPublicKey bool) (err lib.ErrorI) {
+func (p *P2P) AddPeer(conn net.Conn, info *lib.PeerInfo, disconnect, strictPublicKey, closeOnFail bool) (err lib.ErrorI) {
 	// create the e2e encrypted connection while establishing a full peer info object
 	connection, err := p.NewConnection(conn)
 	if err != nil {
 		return err
 	}
-	// always in case of error close connection
-	// first we need to check the error on creation to prevent panics
-	defer func() {
-		if err != nil {
-			connection.Stop()
-		}
-	}()
+	// in case of error close connection when flag is on
+	// not use closeOnFail if using retries
+	if closeOnFail {
```

Comment:
couldn't this be also done with sync.Once?

Reply:
What we want is to not close it at all in some cases.